File: app/src/main/java/com/trackstudio/rfidmanager/MainActivity.java
package com.trackstudio.rfidmanager;

import android.content.SharedPreferences;
import android.media.AudioAttributes;
import android.media.AudioManager;
import android.media.SoundPool;
import android.os.Bundle;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.EditText;
import android.widget.HorizontalScrollView;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

import com.google.android.material.card.MaterialCardView;
import com.google.android.material.slider.Slider;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Locale;

import com.rscja.barcode.BarcodeDecoder;
import com.rscja.barcode.BarcodeFactory;
import com.rscja.barcode.BarcodeUtility;
import com.rscja.deviceapi.RFIDWithUHFUART;
import com.rscja.deviceapi.entity.BarcodeEntity;
import com.rscja.deviceapi.interfaces.ConnectionStatus;
import com.rscja.deviceapi.interfaces.IUHF;

public class MainActivity extends AppCompatActivity {

    private static final String PREFS_NAME = "uhf_settings";
    private static final String PREF_FREQ_MODE = "freq_mode";
    private static final String PREF_POWER = "power";
    private static final int DEFAULT_FREQ_MODE = 2; // Europe 865-868 MHz
    private static final int DEFAULT_POWER = 10;

    private EditText editCode;
    private TextView tvLog;
    private TextView tvStatus;
    private TextView tvPowerValue;
    private ImageView settingsChevron;
    private View settingsBody;
    private AutoCompleteTextView spinnerFreqMode;
    private Slider sliderPower;
    private LinearLayout historyContainer;
    private final LinkedList<String> logEntries = new LinkedList<>();

    private BarcodeDecoder barcodeDecoder;
    private RFIDWithUHFUART mReader;
    private SharedPreferences prefs;
    private boolean settingsExpanded = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        prefs = getSharedPreferences(PREFS_NAME, MODE_PRIVATE);

        editCode = findViewById(R.id.editCode);
        tvLog = findViewById(R.id.tvLog);
        tvStatus = findViewById(R.id.tvStatus);
        tvPowerValue = findViewById(R.id.tvPowerValue);
        settingsChevron = findViewById(R.id.settingsChevron);
        settingsBody = findViewById(R.id.settingsBody);
        spinnerFreqMode = findViewById(R.id.spinnerFreqMode);
        sliderPower = findViewById(R.id.sliderPower);
        historyContainer = findViewById(R.id.historyContainer);

        initSound();
        setupKeyboard();
        setupSettingsToggle();
        setupFrequencyDropdown();
        setupPowerSlider();

        findViewById(R.id.btnWriteRFID).setOnClickListener(v -> performWriteRFID());
        findViewById(R.id.btnReconnect).setOnClickListener(v -> reconnectUhf());

        try {
            barcodeDecoder = BarcodeFactory.getInstance().getBarcodeDecoder();
            mReader = RFIDWithUHFUART.getInstance();

            editCode.setText(R.string.example_code);

            // init barcode scanner
            installBarcodeDecoderCallback();

            // init uhf module
            mReader.init(this);
            if (mReader.getConnectStatus() == ConnectionStatus.CONNECTED) {
                appendLog("Connected");
                applySavedSettings();
                updateSettingsFromReader();
            } else {
                appendLog("Cannot connect Chainway UHF module");
            }

        } catch (Exception ex) {
            appendLog("Init error: " + ex);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        releaseSoundPool();
        if (mReader != null) {
            mReader.free();
        }
        if (barcodeDecoder != null) {
            barcodeDecoder.close();
        }
    }

    // --- Saved settings ---

    private void applySavedSettings() {
        int freqMode = prefs.getInt(PREF_FREQ_MODE, DEFAULT_FREQ_MODE);
        int power = prefs.getInt(PREF_POWER, DEFAULT_POWER);

        mReader.setFrequencyMode(freqMode);
        mReader.setPower(power);
        appendLog("Freq: " + freqMode + ", Power: " + power + " dBm");
    }

    private void saveFreqMode(int mode) {
        prefs.edit().putInt(PREF_FREQ_MODE, mode).apply();
    }

    private void savePower(int power) {
        prefs.edit().putInt(PREF_POWER, power).apply();
    }

    // --- Keyboard fix for Chainway devices ---

    private void showKeyboard() {
        InputMethodManager imm = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);
        editCode.post(() -> {
            editCode.requestFocus();
            imm.showSoftInput(editCode, 0);
        });
    }

    private void setupKeyboard() {
        editCode.setOnFocusChangeListener((v, hasFocus) -> {
            if (hasFocus) showKeyboard();
        });
        editCode.setOnClickListener(v -> showKeyboard());
    }

    // --- Settings panel ---

    private void setupSettingsToggle() {
        findViewById(R.id.settingsHeader).setOnClickListener(v -> {
            settingsExpanded = !settingsExpanded;
            settingsBody.setVisibility(settingsExpanded ? View.VISIBLE : View.GONE);
            settingsChevron.animate().rotation(settingsExpanded ? 180f : 0f).setDuration(200).start();
        });
    }

    private void setupFrequencyDropdown() {
        String[] freqModes = getResources().getStringArray(R.array.frequency_modes);
        ArrayAdapter<String> adapter = new ArrayAdapter<>(this,
                android.R.layout.simple_dropdown_item_1line, freqModes);
        spinnerFreqMode.setAdapter(adapter);
        spinnerFreqMode.setOnItemClickListener((parent, view, position, id) -> {
            if (mReader != null) {
                boolean ok = mReader.setFrequencyMode(position);
                appendLog("Set freq mode: " + freqModes[position] + (ok ? " OK" : " FAILED"));
                if (ok) {
                    saveFreqMode(position);
                }
            }
        });
    }

    private void setupPowerSlider() {
        sliderPower.addOnChangeListener((slider, value, fromUser) -> {
            tvPowerValue.setText(getString(R.string.unit_dbm, (int) value));
            if (fromUser && mReader != null) {
                boolean ok = mReader.setPower((int) value);
                appendLog("Set power: " + (int) value + " dBm" + (ok ? "" : " FAILED"));
                if (ok) {
                    savePower((int) value);
                }
            }
        });
    }

    private void updateSettingsFromReader() {
        if (mReader == null) return;

        tvStatus.setText("Connected");
        tvStatus.setTextColor(getColor(R.color.md_theme_primary));

        // sync frequency dropdown
        int freqMode = mReader.getFrequencyMode();
        String[] freqModes = getResources().getStringArray(R.array.frequency_modes);
        if (freqMode >= 0 && freqMode < freqModes.length) {
            spinnerFreqMode.setText(freqModes[freqMode], false);
        } else {
            spinnerFreqMode.setText("Unknown (" + freqMode + ")", false);
        }

        // sync power slider
        int power = mReader.getPower();
        if (power >= 5 && power <= 30) {
            sliderPower.setValue(power);
        }
        tvPowerValue.setText(getString(R.string.unit_dbm, power));
    }

    // --- Reconnect ---

    private void reconnectUhf() {
        View btnReconnect = findViewById(R.id.btnReconnect);
        btnReconnect.setEnabled(false);
        appendLog("Reconnecting UHF module...");

        new Thread(() -> {
            try {
                if (mReader != null) {
                    mReader.free();
                }
                mReader = RFIDWithUHFUART.getInstance();
                mReader.init(MainActivity.this);

                runOnUiThread(() -> {
                    if (mReader.getConnectStatus() == ConnectionStatus.CONNECTED) {
                        appendLog("Reconnected successfully");
                        applySavedSettings();
                        updateSettingsFromReader();
                    } else {
                        appendLog("Reconnect failed");
                        tvStatus.setText("Disconnected");
                        tvStatus.setTextColor(getColor(R.color.md_theme_error));
                    }
                    btnReconnect.setEnabled(true);
                });
            } catch (Exception ex) {
                runOnUiThread(() -> {
                    appendLog("Reconnect error: " + ex);
                    btnReconnect.setEnabled(true);
                });
            }
        }).start();
    }

    // --- RFID write ---

    private void performWriteRFID() {
        String code = editCode.getText().toString().trim();
        if (code.isEmpty()) {
            appendLog("Code is empty");
            playSound(2);
            return;
        }

        // Validate hex format
        if (!code.matches("^[0-9a-fA-F]+$")) {
            appendLog("Error: \"" + code + "\" is not a valid hex string");
            appendLog("RFID tags only support hex (0-9, A-F)");
            addHistoryTag(code, false);
            playSound(2);
            return;
        }

        new Thread(() -> {
            // default epc size
            int epcSize = 8;
            // remember power
            int power = mReader.getPower();
            mReader.setPower(10); // low range

            try {
                // Guess epc size
                // try to read 8 first
                String currentData = mReader.readData("00000000", IUHF.Bank_EPC, 2, 8);
                if (currentData == null) {
                    // if error - try to read 6
                    currentData = mReader.readData("00000000", IUHF.Bank_EPC, 2, 6);
                    // if ok - fix epcSize
                    if (currentData != null) {
                        epcSize = 6;
                        appendLog("EPC size = 6 words");
                    }
                }

                if (currentData != null) {
                    appendLog("Current RFID tag data (before write): " + currentData);
                } else {
                    String errMessage = ErrorCodeManager.getMessage(mReader.getErrCode());
                    appendLog("Read error: " + errMessage);
                }

                // Check if code fits in detected EPC size
                if (code.length() > 4 * epcSize) {
                    appendLog("Error: Code too long (" + code.length() + " hex chars)");
                    appendLog("Maximum allowed for this tag: " + (4 * epcSize));
                    addHistoryTag(code, false);
                    playSound(2);
                    return;
                }

                // need to write full data
                String format = "%-" + 4 * epcSize + "s"; // %-32s for 8 words, %-24s for 6 words
                String dataToWrite = String.format(format, code).replace(' ', '0');
                boolean result = mReader.writeData("00000000", IUHF.Bank_EPC, 2, epcSize, dataToWrite);
                if (!result) {
                    appendLog("Write error: " + ErrorCodeManager.getMessage(mReader.getErrCode()));
                    addHistoryTag(code, false);
                    playSound(2);
                } else {
                    // Read to re-check
                    String data = mReader.readData("00000000", IUHF.Bank_EPC, 2, epcSize);

                    // Compare
                    if (data != null && data.toLowerCase().startsWith(code.toLowerCase())) {
                        appendLog("SUCCESS " + code.toLowerCase());
                        addHistoryTag(code, true);
                        playSound(1);
                    } else {
                        appendLog("We wrote " + dataToWrite + " but read " + data + " instead");
                        addHistoryTag(code, false);
                        playSound(2);
                    }
                }
            } finally {
                // restore power
                mReader.setPower(power);
            }
        }).start();
    }

    // --- Logging ---

    public void appendLog(String message) {
        runOnUiThread(() -> {
            String timestamp = new SimpleDateFormat("HH:mm:ss", Locale.getDefault()).format(new Date());
            logEntries.addFirst(timestamp + " - " + message);
            if (logEntries.size() > 30) {
                logEntries.removeLast();
            }

            StringBuilder sb = new StringBuilder();
            for (String entry : logEntries) {
                sb.append(entry).append("\n");
            }
            tvLog.setText(sb.toString());

            ScrollView scrollView = findViewById(R.id.log_scroll_view);
            scrollView.post(() -> scrollView.fullScroll(View.FOCUS_UP));
        });
    }

    private void addHistoryTag(String code, boolean success) {
        runOnUiThread(() -> {
            View tagView = getLayoutInflater().inflate(R.layout.history_tag, historyContainer, false);
            TextView tv = tagView.findViewById(R.id.tagText);
            MaterialCardView card = (MaterialCardView) tagView;

            tv.setText(code);
            int colorRes = success ? R.color.md_theme_primary : R.color.md_theme_error;
            card.setCardBackgroundColor(ContextCompat.getColor(this, colorRes));

            // Add to the beginning (left side)
            historyContainer.addView(tagView, 0);

            // Auto-scroll to start (left)
            HorizontalScrollView scroll = findViewById(R.id.historyScroll);
            scroll.post(() -> scroll.fullScroll(View.FOCUS_LEFT));
        });
    }

    // --- Barcode scanner ---

    private void installBarcodeDecoderCallback() {
        if (barcodeDecoder.open(this)) {
            BarcodeUtility.getInstance().enablePlayFailureSound(this, true);
            barcodeDecoder.setDecodeCallback((BarcodeEntity barcodeEntity) -> {
                runOnUiThread(() -> {
                    if (barcodeEntity.getResultCode() == BarcodeDecoder.DECODE_SUCCESS) {
                        editCode.setText(barcodeEntity.getBarcodeData());
                        appendLog("Barcode: " + barcodeEntity.getBarcodeData());
                        performWriteRFID();
                    } else {
                        editCode.setText("");
                        appendLog("Error: " + barcodeEntity.getErrCode());
                    }
                });
            });
        } else {
            appendLog("Cannot connect Chainway barcode scanner");
        }
    }

    // --- Sound ---

    private final HashMap<Integer, Integer> soundMap = new HashMap<>();
    private SoundPool soundPool;
    private float volumnRatio;
    private AudioManager am;

    private void initSound() {
        AudioAttributes audioAttributes = new AudioAttributes.Builder()
                .setUsage(AudioAttributes.USAGE_ASSISTANCE_SONIFICATION)
                .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                .build();
        soundPool = new SoundPool.Builder()
                .setMaxStreams(10)
                .setAudioAttributes(audioAttributes)
                .build();
        soundMap.put(1, soundPool.load(this, R.raw.barcodebeep, 1));
        soundMap.put(2, soundPool.load(this, R.raw.serror, 1));
        am = (AudioManager) this.getSystemService(AUDIO_SERVICE);
    }

    private void releaseSoundPool() {
        if (soundPool != null) {
            soundPool.release();
            soundPool = null;
        }
    }

    /**
     * Play notification sound
     *
     * @param id 1 = success, 2 = error
     */
    public void playSound(int id) {
        Integer soundId = soundMap.get(id);
        if (soundId == null || soundPool == null) {
            return;
        }
        float audioMaxVolume = am.getStreamMaxVolume(AudioManager.STREAM_MUSIC);
        float audioCurrentVolume = am.getStreamVolume(AudioManager.STREAM_MUSIC);
        volumnRatio = audioCurrentVolume / audioMaxVolume;
        try {
            soundPool.play(soundId, volumnRatio, volumnRatio, 1, 0, 1);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/com/trackstudio/rfidmanager/MainActivity.kt
package com.trackstudio.rfidmanager

import android.content.Context
import android.content.SharedPreferences
import android.media.AudioAttributes
import android.media.AudioManager
import android.media.SoundPool
import android.os.Bundle
import android.view.KeyEvent
import android.view.View
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.NavHostFragment
import androidx.navigation.ui.setupWithNavController
import com.trackstudio.rfidmanager.databinding.ActivityMainBinding
import com.rscja.barcode.BarcodeDecoder
import com.rscja.barcode.BarcodeFactory
import com.rscja.barcode.BarcodeUtility
import com.rscja.deviceapi.RFIDWithUHFUART
import com.rscja.deviceapi.interfaces.IUHF
import com.rscja.deviceapi.interfaces.IUHFLocationCallback
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.util.concurrent.atomic.AtomicBoolean
import java.text.SimpleDateFormat
import java.util.*

class MainActivity : AppCompatActivity() {

    companion object {
        private const val PREFS_NAME = "uhf_settings"
        private const val PREF_FREQ_MODE = "freq_mode_id" // Use ID instead of index
        private const val PREF_WRITE_POWER = "write_power"
        private const val PREF_RADAR_POWER = "radar_power"
        private const val DEFAULT_WRITE_POWER = 10
        private const val DEFAULT_RADAR_POWER = 30
        
        // Frequency Mode IDs from Vendor SDK
        const val MODE_CHINA_840_845 = 0x01
        const val MODE_CHINA_920_925 = 0x02
        const val MODE_EUROPE_865_868 = 0x04
        const val MODE_USA_902_928 = 0x08
        const val MODE_KOREA = 0x16
    }

    private lateinit var binding: ActivityMainBinding
    private val viewModel: SharedViewModel by viewModels()
    
    private var barcodeDecoder: BarcodeDecoder? = null
    private var mReader: RFIDWithUHFUART? = null
    private lateinit var prefs: SharedPreferences
    private val logEntries = LinkedList<String>()
    private val historyEntries = mutableListOf<Pair<String, Boolean>>()
    private var originalRadarPower: Int? = null
    private val uhfMutex = Mutex()

    // Sound
    private val soundMap = HashMap<Int, Int>()
    private var soundPool: SoundPool? = null
    private lateinit var am: AudioManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        prefs = getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

        val navHostFragment = supportFragmentManager.findFragmentById(R.id.nav_host_fragment) as NavHostFragment
        val navController = navHostFragment.navController
        binding.bottomNavigation?.setupWithNavController(navController)

        initSound()
        setupStatusObserver()
        initHardware()
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        // Side trigger is not intercepted; SDK/location mode owns UHF work.
        // Let system/vendor key handling continue normally.
        return super.onKeyDown(keyCode, event)
    }

    override fun onKeyUp(keyCode: Int, event: KeyEvent): Boolean {
        return super.onKeyUp(keyCode, event)
    }

    private fun setupStatusObserver() {
        viewModel.statusText.observe(this) { status ->
            binding.tvGlobalStatus?.text = status
        }
        viewModel.isConnected.observe(this) { connected ->
            val color = if (connected) R.color.md_theme_primary else R.color.md_theme_error
            binding.ivStatusDot?.setColorFilter(ContextCompat.getColor(this, color))
        }
    }

    private fun initHardware() {
        try {
            barcodeDecoder = BarcodeFactory.getInstance().barcodeDecoder
            installBarcodeDecoderCallback()
            appendLog("Initializing UHF module...")
            lifecycleScope.launch(Dispatchers.IO) {
                completeUhfInitialization("startup")
            }
        } catch (ex: Exception) {
            appendLog("Init error: $ex")
        }
    }

    override fun onDestroy() {
        CoroutineScope(Dispatchers.IO).launch {
            uhfMutex.withLock {
                mReader?.let { reader ->
                    runCatching {
                        if (isLocationRunning.get()) {
                            reader.stopLocation()
                            isLocationRunning.set(false)
                        }
                        reader.stopInventory()
                        reader.free()
                    }.onFailure { appendLog("UHF destroy cleanup error: $it") }
                    mReader = null
                }
            }
        }
        releaseSoundPool()
        barcodeDecoder?.close()
        super.onDestroy()
    }

    private suspend fun applySavedSettings() {
        uhfMutex.withLock {
            if (prefs.contains(PREF_FREQ_MODE)) {
                val freqMode = prefs.getInt(PREF_FREQ_MODE, -1)
                if (freqMode != -1) {
                    mReader?.setFrequencyMode(freqMode)
                }
            }
        }
    }

    fun saveFreqMode(modeId: Int) {
        prefs.edit().putInt(PREF_FREQ_MODE, modeId).apply()
        lifecycleScope.launch(Dispatchers.IO) {
            uhfMutex.withLock {
                mReader?.setFrequencyMode(modeId)
            }
        }
    }

    fun getWritePower() = prefs.getInt(PREF_WRITE_POWER, DEFAULT_WRITE_POWER)
    fun saveWritePower(power: Int) {
        prefs.edit().putInt(PREF_WRITE_POWER, power).apply()
    }

    fun getRadarPower() = prefs.getInt(PREF_RADAR_POWER, DEFAULT_RADAR_POWER)
    fun saveRadarPower(power: Int) {
        prefs.edit().putInt(PREF_RADAR_POWER, power).apply()
        setDynamicRadarPower(power)
    }

    fun openBarcode() {
        if (barcodeDecoder?.open(this) == true) {
            BarcodeUtility.getInstance().enablePlayFailureSound(this, true)
            appendLog("Barcode scanner enabled")
        }
    }

    fun closeBarcode() {
        barcodeDecoder?.close()
        appendLog("Barcode scanner disabled (Radar mode)")
    }

    fun reconnectUhf() {
        appendLog("Reconnecting UHF module...")
        lifecycleScope.launch(Dispatchers.IO) {
            try {
                completeUhfInitialization("reconnect")
            } catch (ex: Exception) {
                appendLog("Reconnect error: $ex")
            }
        }
    }

    private suspend fun completeUhfInitialization(reason: String) {
        val result = uhfMutex.withLock { initializeUhfLocked(reason) }
        viewModel.setConnectionStatus(result.connected)

        if (result.connected) {
            val deviceFreq = uhfMutex.withLock { mReader?.frequencyMode ?: -1 }
            appendLog("UHF $reason final success: initPath=${result.attempts.lastOrNull()?.initPath}, connectStatus=${result.attempts.lastOrNull()?.connectStatus}, isPowerOn=${result.attempts.lastOrNull()?.isPowerOn}, errCode=${result.attempts.lastOrNull()?.errCode}")
            appendLog("Connected. Hardware region: 0x${Integer.toHexString(deviceFreq)}")
            if (!prefs.contains(PREF_FREQ_MODE) && deviceFreq != -1) {
                prefs.edit().putInt(PREF_FREQ_MODE, deviceFreq).apply()
            }
            applySavedSettings()
            withContext(Dispatchers.Main) { syncSettingsToFragment() }
        } else {
            val lastAttempt = result.attempts.lastOrNull()
            appendLog("UHF $reason final failure: attempts=${result.attempts.size}, initPath=${lastAttempt?.initPath}, connectStatus=${lastAttempt?.connectStatus}, isPowerOn=${lastAttempt?.isPowerOn}, errCode=${lastAttempt?.errCode}")
        }
    }

    private suspend fun initializeUhfLocked(reason: String): UhfInitResult {
        val previousReader = mReader?.let { ChainwayUhfReaderAdapter(it, this@MainActivity, ChainwayUhfInitStrategy.SYSTEM_POWER_CONTEXT_INIT) }
        val controller = UhfConnectionController(
            scannerCleaner = ChainwayScannerCleaner(this@MainActivity),
            readerFactory = ChainwayUhfReaderFactory(this@MainActivity),
            logHandler = { appendUhfAttemptLog(reason, it) }
        )
        val result = controller.initialize(previousReader)
        mReader = (result.reader as? ChainwayUhfReaderAdapter)?.reader
        return result
    }

    private fun appendUhfAttemptLog(reason: String, attemptLog: UhfInitAttemptLog) {
        appendLog(
                "UHF $reason attempt ${attemptLog.attempt}: " +
                "isUhfWorking=${attemptLog.isUhfWorking}, " +
                "scannerEnable=${attemptLog.scannerEnableResult}, " +
                "scannerStop=${attemptLog.scannerStopResult}, " +
                "scannerDisable=${attemptLog.scannerDisableResult}, " +
                "oldFree=${attemptLog.previousReaderFreeResult ?: "not-needed"}, " +
                "initPath=${attemptLog.initPath}, " +
                "initResult=${attemptLog.initResult}, " +
                "connectStatus=${attemptLog.connectStatus}, " +
                "isPowerOn=${attemptLog.isPowerOn}, " +
                "errCode=${attemptLog.errCode}, " +
                "result=${attemptLog.result}, " +
                "nextRetryDelayMs=${attemptLog.nextRetryDelayMs}"
        )
        if (attemptLog.cleanupExceptions.isNotEmpty()) {
            appendLog("UHF $reason attempt ${attemptLog.attempt} non-fatal cleanup exceptions: ${attemptLog.cleanupExceptions.joinToString()}")
        }
    }

    fun performWriteRFID(code: String) {
        val initialTarget = try {
            EpcTargetNormalizer.normalize(code)
        } catch (ex: IllegalArgumentException) {
            appendLog("Write rejected: ${ex.message}")
            addHistoryTag(code, false)
            playSound(2)
            return
        }

        lifecycleScope.launch(Dispatchers.IO) {
            val outcome = uhfMutex.withLock {
                val reader = mReader ?: return@launch
                var epcSize = 8
                val originalPower = reader.power
                val targetWritePower = getWritePower()
                reader.setPower(targetWritePower)

                try {
                    var currentData = reader.readData("00000000", IUHF.Bank_EPC, 2, 8)
                    if (currentData == null) {
                        currentData = reader.readData("00000000", IUHF.Bank_EPC, 2, 6)
                        if (currentData != null) epcSize = 6
                    }

                    val target = try {
                        EpcTargetNormalizer.normalize(initialTarget.label, epcSize)
                    } catch (ex: IllegalArgumentException) {
                        return@withLock WriteOutcome("Write rejected: ${ex.message}", initialTarget.label, false)
                    }

                    val result = reader.writeData("00000000", IUHF.Bank_EPC, 2, epcSize, target.paddedHex)
                    if (!result) {
                        WriteOutcome("Write error: ${ErrorCodeManager.getMessage(reader.errCode)}", target.label, false)
                    } else {
                        val verifyData = reader.readData("00000000", IUHF.Bank_EPC, 2, epcSize)
                        if (verifyData != null && verifyData.startsWith(target.label, ignoreCase = true)) {
                            WriteOutcome("SUCCESS ${target.label.lowercase()}", target.label, true)
                        } else {
                            WriteOutcome("Verification failed", target.label, false)
                        }
                    }
                } finally {
                    reader.setPower(originalPower)
                }
            }

            withContext(Dispatchers.Main) {
                appendLog(outcome.message)
                addHistoryTag(outcome.label, outcome.success)
                playSound(if (outcome.success) 1 else 2)
            }
        }
    }

    private data class WriteOutcome(
        val message: String,
        val label: String,
        val success: Boolean
    )

    // --- Radar ---
    private val isLocationRunning = AtomicBoolean(false)

    /** Updates idle locate power without interrupting an active Chainway location session. */
    fun setDynamicRadarPower(power: Int) {
        lifecycleScope.launch(Dispatchers.IO) {
            uhfMutex.withLock {
                val reader = mReader ?: return@withLock
                if (!isLocationRunning.get()) {
                    reader.setPower(power)
                }
            }
        }
    }

    fun startRadar(
        targetMask: String,
        onStartResult: (Boolean) -> Unit,
        onLocationValue: (Int, Boolean) -> Unit
    ) {
        val target = try {
            EpcTargetNormalizer.normalize(targetMask)
        } catch (ex: IllegalArgumentException) {
            appendLog("Radar rejected: ${ex.message}")
            onStartResult(false)
            return
        }

        lifecycleScope.launch(Dispatchers.IO) {
            val started = uhfMutex.withLock {
                val reader = mReader ?: return@withLock false
                val targetRadarPower = getRadarPower()
                originalRadarPower = reader.power
                reader.setPower(targetRadarPower)

                val locationStarted = reader.startLocation(
                    this@MainActivity,
                    target.label,
                    IUHF.Bank_EPC,
                    32,
                    object : IUHFLocationCallback {
                        override fun getLocationValue(value: Int, found: Boolean) {
                            lifecycleScope.launch(Dispatchers.Main) {
                                onLocationValue(value.coerceIn(0, 100), found)
                            }
                        }
                    }
                )

                if (locationStarted) {
                    isLocationRunning.set(true)
                } else {
                    val savedPower = originalRadarPower
                    originalRadarPower = null
                    savedPower?.let { reader.setPower(it) }
                }

                locationStarted
            }

            if (started) {
                withContext(Dispatchers.Main) {
                    appendLog("Radar: Locate started for *${target.label}*")
                    onStartResult(true)
                }
            } else {
                withContext(Dispatchers.Main) {
                    appendLog("Radar Error: Failed to start Locate mode")
                    onStartResult(false)
                }
            }
        }
    }

    fun stopRadar() {
        lifecycleScope.launch(Dispatchers.IO) {
            uhfMutex.withLock {
                val reader = mReader ?: return@withLock
                isLocationRunning.set(false)
                reader.stopLocation()
                originalRadarPower?.let { reader.setPower(it) }
                originalRadarPower = null
            }
        }
    }

    // --- Fragment Sync ---

    fun updateWriteFragmentUI(fragment: WriteFragment) {
        historyEntries.forEach { fragment.addHistoryTag(it.first, it.second) }
    }

    fun updateLogFragmentUI(fragment: LogFragment) {
        fragment.setLogs(logEntries)
    }

    fun updateSettingsFragmentUI(fragment: SettingsFragment) {
        val freqMode = prefs.getInt(PREF_FREQ_MODE, -1)
        fragment.updateUI(freqMode)
    }

    private fun syncSettingsToFragment() {
        val navHost = supportFragmentManager.findFragmentById(R.id.nav_host_fragment) as NavHostFragment
        val fragment = navHost.childFragmentManager.fragments.firstOrNull() as? SettingsFragment
        fragment?.let { updateSettingsFragmentUI(it) }
    }

    fun appendLog(message: String) {
        lifecycleScope.launch(Dispatchers.Main) {
            val timestamp = SimpleDateFormat("HH:mm:ss", Locale.getDefault()).format(Date())
            logEntries.addFirst("$timestamp - $message")
            if (logEntries.size > 30) logEntries.removeAt(logEntries.size - 1)
            
            val navHost = supportFragmentManager.findFragmentById(R.id.nav_host_fragment) as NavHostFragment
            val fragment = navHost.childFragmentManager.fragments.firstOrNull() as? LogFragment
            fragment?.setLogs(logEntries)
        }
    }

    private fun addHistoryTag(code: String, success: Boolean) {
        lifecycleScope.launch(Dispatchers.Main) {
            historyEntries.add(0, code to success)
            if (historyEntries.size > 20) historyEntries.removeAt(historyEntries.lastIndex)
            
            val navHost = supportFragmentManager.findFragmentById(R.id.nav_host_fragment) as NavHostFragment
            val fragment = navHost.childFragmentManager.fragments.firstOrNull() as? WriteFragment
            fragment?.addHistoryTag(code, success)
        }
    }

    private fun installBarcodeDecoderCallback() {
        barcodeDecoder?.let { decoder ->
            // Don't open here, fragments will manage lifecycle
            decoder.setDecodeCallback { barcodeEntity ->
                lifecycleScope.launch(Dispatchers.Main) {
                    if (barcodeEntity.resultCode == BarcodeDecoder.DECODE_SUCCESS) {
                        val navHost = supportFragmentManager.findFragmentById(R.id.nav_host_fragment) as NavHostFragment
                        val fragment = navHost.childFragmentManager.fragments.firstOrNull() as? WriteFragment
                        fragment?.setBarcodeData(barcodeEntity.barcodeData)
                        performWriteRFID(barcodeEntity.barcodeData)
                    }
                }
            }
        }
    }

    // --- Utils ---
    private fun initSound() {
        val audioAttributes = AudioAttributes.Builder()
            .setUsage(AudioAttributes.USAGE_ASSISTANCE_SONIFICATION)
            .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION).build()
        soundPool = SoundPool.Builder().setMaxStreams(10).setAudioAttributes(audioAttributes).build().apply {
            soundMap[1] = load(this@MainActivity, R.raw.barcodebeep, 1)
            soundMap[2] = load(this@MainActivity, R.raw.serror, 1)
        }
        am = getSystemService(Context.AUDIO_SERVICE) as AudioManager
    }

    private fun releaseSoundPool() {
        soundPool?.release()
        soundPool = null
    }

    fun playSound(id: Int, rate: Float = 1f) {
        val soundId = soundMap[id] ?: return
        val pool = soundPool ?: return
        val volume = am.getStreamVolume(AudioManager.STREAM_MUSIC).toFloat() / am.getStreamMaxVolume(AudioManager.STREAM_MUSIC)
        try {
            pool.play(soundId, volume, volume, 1, 0, rate.coerceIn(0.5f, 2.0f))
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}
